cmd/zc_traverser_list.go

```diff
@@ -73,15 +73,12 @@ func (l *listTraverser) traverse(processor objectProcessor, filters []objectFilt
 		// case 2: child2 is a directory, and it has items under it such as child2/grandchild1
 		//         the relative path returned by the child traverser would be "grandchild1"
 		//         it should be "child2/grandchild1" instead
-		preProcessor := func(object storedObject) error {
+		childPreProcessor := func(object *storedObject) {
 			object.relativePath = common.GenerateFullPath(childPath, object.relativePath)
-			return processor(object)
 		}
+		preProcessorForThisChild := preprocessor.FollowedBy(childPreProcessor) // always chain pre-processors, never replace them (this is so we avoid making any assumptions about whether an old processor actually exists)
```

Comment:
I'm not entirely sure how I feel about wrapping function after function in generated functions. This may well actually end up making debugging harder. In my experience, the debugger generally won't jump into them unless I add an explicit breakpoint within the generated functions. The problem is, I'm also not entirely sure what the best plan of action is to remedy this.

Reply:
I think the other way to do it would be to define objectMorpher as an interface, not a func, and to create objects that implemented that interface. E.g. a "container setter" and a "path prepender". And have newStoredObject take an array of objectMorphers, and FollowedBy just append to the array.

I think that would all work, and the code would mostly look about the same. And as you say, debuggability would be improved.

I'll put a TODO in the code to say that we should consider that.